File: C++/026.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include "Utils.h"

using namespace std;

int recurLen(int d) {
  vector<int> v;
  int count = 0;
  int n = 1;
  // written division
  while (1) {
    n *= 10;
    n = n%d;
    auto f = find(v.begin(), v.end(), n);
    if (f != v.end()) {
      return count-(f-v.begin());
    }
    v.push_back(n);
    count++;
  }
}

int main() {
  int limit = 1000;
  int res = 0;
  int resIndex = 0;
  bool primes[limit];
  sieve(limit, primes);

  for (int i=limit-1; i>=0; i--) {
    if (primes[i] == true) { // if isprime
      int len = recurLen(i);
      if (len > res) {
        res = len;
        resIndex = i;
      }
      else
        break;
    }
  }
  cout << resIndex;

}

File: Python/026.py
from Utils import sieve_of_eratosthenes

# no need for euler totient function, because only small numbers
def recur_len(d):
    if d % 2 != 0 and d % 5 != 0 and d != 0 and d != 1:
        n = 1
        # every fraction 1/d can be written as x/(10**n - 1) where n is the length 
        # of the recurring cycle
        while (10**n - 1) % d != 0:
            n += 1
        return n

res = 0
res_index = 0
sieve = sieve_of_eratosthenes(1000,lower_limit=6)

for i in reversed(sieve):
    length = recur_len(i)
    if length > res:
        res = length
        res_index = i
    if i < res:
        break

print(res_index)